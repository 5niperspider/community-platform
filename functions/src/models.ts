import * as functions from 'firebase-functions'

// Models can be imported from the main package for use here
// NOTE 1 - this requires adjustment main src in package.json
// NOTE 2 - shorthand @OAModels notation defined in tsconfig
export * from '../../src/models/common.models'
export * from '../../src/models/user.models'
export * from '../../src/models/howto.models'
export * from '../../src/models/events.models'

export type IDBDocChange = functions.Change<FirebaseFirestore.DocumentSnapshot>

// TODO - handle import from src/models (currently breaks ts setup)
// but possibly fixed by backend improvements pr
// NOTE - types import fine as skipLib=true, but consts have issue
const e = process.env
export const DB_PREFIX = e.REACT_APP_DB_PREFIX ? e.REACT_APP_DB_PREFIX : 'v3_'
