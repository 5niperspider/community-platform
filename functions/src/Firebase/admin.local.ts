/***********************************************************
When serving locally use service account info provided

***********************************************************/
import * as admin from 'firebase-admin'
const FIREBASE_CONFIG = JSON.parse(
  process.env.FIREBASE_CONFIG,
) as IFirebaseConfig

const serviceAccount = require('../config/serviceAccount.json')

admin.initializeApp({
  credential: admin.credential.cert({
    clientEmail: serviceAccount.client_email,
    privateKey: serviceAccount.private_key,
    projectId: serviceAccount.project_id,
  }),
  databaseURL: FIREBASE_CONFIG.databaseURL,
})
export const firebaseAdmin = admin

// this is automatically populated by firebase
interface IFirebaseConfig {
  databaseURL: string
  storageBucket: string
  projectId: string
}
