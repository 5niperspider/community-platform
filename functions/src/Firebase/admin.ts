/***********************************************************
When running on live the config below pulls all data from 
environment variables.
***********************************************************/
import * as admin from 'firebase-admin'
import { SERVICE_ACCOUNT_CONFIG } from '../config/config'

const serviceAccount = SERVICE_ACCOUNT_CONFIG
const FIREBASE_CONFIG = JSON.parse(
  process.env.FIREBASE_CONFIG,
) as IFirebaseConfig

if (serviceAccount) {
  const cert = {
    clientEmail: serviceAccount.client_email,
    privateKey: serviceAccount.private_key,
    projectId: serviceAccount.project_id,
  }
  admin.initializeApp({
    credential: admin.credential.cert(cert),
    databaseURL: FIREBASE_CONFIG.databaseURL,
    projectId: FIREBASE_CONFIG.projectId,
  })
}
else {
  admin.initializeApp()
}


export const firebaseAdmin = admin

// this is automatically populated by firebase
export interface IFirebaseConfig {
  databaseURL: string
  storageBucket: string
  projectId: string
}
