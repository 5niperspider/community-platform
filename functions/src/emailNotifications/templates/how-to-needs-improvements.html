{{#> layout}}

<p>Hey {{user.displayName}},</p>

<p>Your How-To, {{howto.title}}, needs more work before we can publish it.</p>

{{#if howto.moderatorFeedback}}
<p>Please review the following & re-submit your How-to:</p>
{{howto.moderatorFeedback}} {{/if}}

<p>We look forward to hearing back from you.</p>

{{/layout}}
