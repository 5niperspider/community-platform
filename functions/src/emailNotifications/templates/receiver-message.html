{{#> layout}}

<p>Hey you got a new message from <strong>{{fromUser}}</strong>:</p>

<p>
  <strong>{{text}}</strong>
</p>

<p>Make sure to reply to their email address {{email}}.</p>

{{/layout}}
