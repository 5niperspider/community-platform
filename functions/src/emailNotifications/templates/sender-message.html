{{#> layout}}

<p>We send the following message to <strong>{{toUserName}}</strong>:</p>

<p>
  {{text}}
</p>

{{/layout}}
