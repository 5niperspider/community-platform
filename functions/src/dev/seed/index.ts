import { Request, Response } from "express"
// (note - typings don't currently exist for firebase-tools: https://github.com/firebase/firebase-tools/issues/2378)
import * as firebase_tools from 'firebase-tools'
import { DB_ENDPOINTS } from "../../models"
import { db } from "../../Firebase/firestoreDB"
import { splitArrayToChunks } from "../../Utils/data.utils"
import { firestore } from "firebase-admin"


/**
 * Script used to generate a clean seed of seed data for use in development
 * 
 * Given a full dump of site data in the emulator, strips away all user docs and revisions
 * for users that have not posted content to mappins, howtos, research or events
 * (basically anywhere their profile might be linked from)
 */
export async function cleanSeedData(req: Request, res: Response) {
    const dbCollections = await db.listCollections()
    const dbEndpoints = dbCollections.map(c => c.id)
    const expectedEndpoints = Object.values(DB_ENDPOINTS)
    // Delete collections not in use
    for (const endpoint of dbEndpoints) {
        if (!expectedEndpoints.includes(endpoint)) {
            await deleteCollectionCLI(endpoint)
        }
    }
    // Get list of users with howtos, mappins or events to retain data
    const keptUsers = {}
    const endpointsToCheck = ['mappins', 'howtos', 'research', 'events'] as const
    for (const endpoint of endpointsToCheck) {
        const mappedEndpoint = DB_ENDPOINTS[endpoint]
        if (mappedEndpoint) {
            const docs = await db.collection(mappedEndpoint).get()
            docs.forEach(doc => {
                switch (endpoint) {
                    case 'mappins':
                        keptUsers[doc.id] = true
                        break;
                    default:
                        const data = doc.data();
                        keptUsers[data._createdBy] = true
                }
            })
        }
    }
    // Delete non-required users in batches 
    const deletedUsers = await deleteQueryDocs(db.collection(DB_ENDPOINTS.users), '[Users Deleted]', (doc) => !keptUsers[doc.data()._id])
    // Delete nested revision docs (TODO - should add filter to ensure doc ref is a subcollection of users)
    const deletedRevisions = await deleteQueryDocs(db.collectionGroup('revisions'), '[Revisions Deleted]', (doc) => !keptUsers[doc.data()._id])
    // Delete nested stats docs
    const deletedStats = await deleteQueryDocs(db.collectionGroup('stats'), '[Stats Deleted]', (doc) => !keptUsers[doc.data()._id])
    res.status(200).send({ keptUsers, deletedUsers, deletedRevisions, deletedStats })
}

async function deleteQueryDocs(
    query: firestore.CollectionGroup<firestore.DocumentData> | firestore.CollectionReference<firestore.DocumentData>,
    logPrefix = '[Deleted]',
    filterFn = (doc: firestore.QueryDocumentSnapshot) => false) {
    const queryResults = await query.get()
    const filteredRefs = queryResults.docs.filter(doc => filterFn(doc)).map(doc => doc.ref)
    const chunks = splitArrayToChunks(filteredRefs, 500)
    let currentChunk = 0
    const totalChunks = chunks.length
    for (const chunk of chunks) {
        const batch = db.batch()
        for (const ref of chunk) {
            batch.delete(ref)
        }
        await batch.commit()
        currentChunk++
        console.log(`${logPrefix} [${currentChunk * 500}] of ${totalChunks * 500}`)
        await _waitForNextTick()
    }
    return filteredRefs
}

/** Simple wrapper function to avoid cascading knock-ons for triggered functions etc. */
function _waitForNextTick(): Promise<void> {
    return new Promise((resolve) => process.nextTick(() => resolve()))
}

/**
 * Use firebase tools to fully delete collection and subcollection from command-line tools
 * https://firebase.google.com/docs/firestore/solutions/delete-collections
 */
async function deleteCollectionCLI(endpoint: string) {
    // Note - whilst we are only operating on the emulator user will probably
    // still need to be logged into firebase to call
    // https://github.com/firebase/firebase-tools/issues/1940
    await firebase_tools.firestore.delete(endpoint, {
        recursive: true,
        yes: true,
    });
}


/**
 * Iterative method to delete all documents in a collection, by reading all documents in batches and deleting
 * Copied from https://github.com/firebase/snippets-node/blob/e5f6214059bdbc63f94ba6600f7f84e96325548d/firestore/main/index.js#L889-L921
 *
 * Note 1 - this is less efficient than the cli method, but also available outside of node environment
 * Note 2 - This fails to delete subcollections, and so more work required to make complete
 */
async function deleteCollectionIteratively(collectionPath: string, batchSize = 500) {
    const collectionRef = db.collection(collectionPath);
    const query = collectionRef.orderBy('__name__').limit(batchSize);
    return new Promise((resolve, reject) => {
        deleteQueryBatch(query, resolve).catch(reject);
    });
}

async function deleteQueryBatch(query, resolve) {
    const snapshot = await query.get();

    const batchSize = snapshot.size;
    if (batchSize === 0) {
        // When there are no documents left, we are done
        resolve();
        return;
    }

    // Delete documents in a batch
    const batch = db.batch();
    snapshot.docs.forEach((doc) => {
        batch.delete(doc.ref);
    });
    await batch.commit();

    // Recurse on the next process tick, to avoid
    // exploding the stack.
    process.nextTick(async () => {
        await deleteQueryBatch(query, resolve);
    });
}
