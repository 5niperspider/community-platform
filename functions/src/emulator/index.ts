import * as functions from 'firebase-functions'
import cors from 'cors'
import express from 'express'
import { PubSub } from '@google-cloud/pubsub'
import { seedDataClean } from './seed/data-clean'
import { seedUsersCreate } from './seed/users-create'

console.log('emulator api ready')

const app = express()
app.use(cors({ origin: true }))

/**
 * Expose an api endpoint available just for use with emulators
 * NOTE - if changing endpoints check emulators workspace in case in use (e.g. called via Dockerfile)
 */
app.use(function (req, res, next) {
  if (process.env.FUNCTIONS_EMULATOR === 'true') next()
  else
    res
      .status(403)
      .send(
        `Emulator api methods can only be accessed in emulator environment [${process.env.FUNCTIONS_EMULATOR}]`,
      )
})

app.get('/', (req, res) => res.status(200).send('Emulator Api Working'))

app.get('/seed-users-create', (req, res) =>
  seedUsersCreate().then((users) =>
    res
      .status(200)
      .send(`Created users: ${users.map((u) => u.uid).join(', ')}`),
  ),
)

app.get('/seed-clean', (req, res) =>
  seedDataClean().then((result) => {
    res.status(200).send(result)
  }),
)

/**
 * As there is currently no UI to publish pubsub messages for testing, create
 * a http wrapper that can be used to call via rest
 * https://github.com/firebase/firebase-tools/issues/2034
 * https://blog.minimacode.com/publish-message-to-pubsub-emulator/
 *
 * This allows pubsub topics to be tested when running in emulators as an api endpoint
 * It includes additional prefix (/topic or /schedule) depending on trigger
 * @example
 * `http://localhost:4002/community-platform-emulated/us-central1/emulator/pubsub/schedule/dailyTasks`
 * `http://localhost:4002/community-platform-emulated/us-central1/emulator/pubsub/topic/test-topic`
 */
app.post('/pubsub/:trigger/:topic', async (req, res) => {
  if (!process.env.PUBSUB_EMULATOR_HOST) {
    functions.logger.error(
      'This function should only run locally in an emulator.',
    )
    res.status(400).end()
  }
  const pubsub = new PubSub()

  const trigger: 'topic' | 'schedule' = req.params.trigger as any
  let topic: string = req.params.topic

  console.log({ trigger, topic })

  if (trigger === 'schedule') {
    // prefix scheduled functions to be picked up by firebase
    // https://github.com/firebase/firebase-tools/pull/2011/files
    topic = `firebase-schedule-${topic}`
  }

  const [topics] = await pubsub.getTopics()
  if (!topics.find((t) => t.name.includes(topic))) {
    await pubsub.createTopic(topic)
  }

  const payload = req.body

  const messageID = await pubsub.topic(topic).publishMessage({ json: payload })
  res.status(201).send({
    success: 'Published to pubsub',
    topic,
    messageID,
    payload,
  })
})

// Increase default timeout to max allowed for longer-running operations
export = functions.runWith({ timeoutSeconds: 540 }).https.onRequest(app as any)
