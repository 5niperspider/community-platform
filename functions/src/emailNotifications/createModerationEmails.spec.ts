import { FirebaseEmulatedTest } from '../test/Firebase/emulator'
import { DB_ENDPOINTS, IMapPin, IUserDB } from '../models'
import { HOW_TO_APPROVAL_SUBJECT, MAP_PIN_APPROVAL_SUBJECT } from './templates'
import { setMockHowto } from '../emulator/seed/content-generate'
import {
  createHowtoModerationEmail,
  createMapPinModerationEmail,
  handleModerationUpdate,
} from './createModerationEmails'

jest.mock('../Firebase/auth', () => ({
  firebaseAuth: {
    getUser: () => ({
      email: 'test@test.com',
    }),
  },
}))

jest.mock('../config/config', () => ({
  CONFIG: {
    deployment: {
      site_url: 'https://community.preciousplastic.com',
    },
  },
}))

const userFactory = (_id: string, user: Partial<IUserDB> = {}): IUserDB =>
  ({
    _id,
    _authID: _id,
    ...user,
  } as IUserDB)

describe('Create howto moderation emails', () => {
  const db = FirebaseEmulatedTest.admin.firestore()

  beforeAll(async () => {
    await FirebaseEmulatedTest.clearFirestoreDB()
    await FirebaseEmulatedTest.seedFirestoreDB('emails')

    await FirebaseEmulatedTest.seedFirestoreDB('users', [
      userFactory('user_1', {
        displayName: 'User 1',
        userName: 'user_1',
        userRoles: ['beta-tester'],
      }),
      userFactory('user_2', {
        displayName: 'User 2',
        userName: 'user_2',
      }),
    ])

    await FirebaseEmulatedTest.seedFirestoreDB('howtos')
  })

  afterAll(async () => {
    await FirebaseEmulatedTest.clearFirestoreDB()
  })

  it('Creates an email for an accepted howto', async () => {
    const howtoApproved = await setMockHowto({ uid: 'user_1' })
    const howtoAwaitingModeration = {
      ...howtoApproved,
      moderation: 'awaiting-moderation',
    }
    const change = FirebaseEmulatedTest.mockFirestoreChangeObject(
      howtoAwaitingModeration,
      howtoApproved,
      'howtos',
      howtoApproved._id,
    )

    await handleModerationUpdate(change, createHowtoModerationEmail)

    // Only one approved howto email should have been created
    const countSnapshot = await db.collection(DB_ENDPOINTS.emails).count().get()
    expect(countSnapshot.data().count).toEqual(1)

    const querySnapshot = await db.collection(DB_ENDPOINTS.emails).get()
    querySnapshot.forEach((doc) => {
      const {
        message: { html, subject },
        to,
      } = doc.data()
      expect(html).toMatchSnapshot()
      expect(subject).toBe(HOW_TO_APPROVAL_SUBJECT)
      // Check that the email contains the correct user name
      expect(html).toContain('Hey User 1')
      // Check that the email contains the correct howto title
      expect(html).toContain(
        'Huzzah! Your How-To Mock Howto has been approved.',
      )
      // Check that the email contains the correct howto link
      expect(html).toContain(
        'https://community.preciousplastic.com/how-to/00_user_1_howto',
      )
      // Check that the email contains the correct PP signoff
      expect(html).toContain('Charlie your Precious Plastic Community Manager')
      expect(to).toBe('test@test.com')
    })
  })

  // Remove this test once released to all users.
  it('Does not creates email for people who are not beta testers', async () => {
    const howtoApproved = await setMockHowto({ uid: 'user_2' })
    const howtoAwaitingModeration = {
      ...howtoApproved,
      moderation: 'awaiting-moderation',
    }
    const change = FirebaseEmulatedTest.mockFirestoreChangeObject(
      howtoAwaitingModeration,
      howtoApproved,
      'howtos',
      howtoApproved._id,
    )

    await handleModerationUpdate(change, createHowtoModerationEmail)

    // No new emails should have been created
    const countSnapshot = await db.collection(DB_ENDPOINTS.emails).count().get()
    expect(countSnapshot.data().count).toEqual(1)
  })

  it('Does not creates email for non-approved howtos', async () => {
    const howtoApproved = await setMockHowto({ uid: 'user_1' })
    const howtoDraft = {
      ...howtoApproved,
      moderation: 'draft',
    }
    const change = FirebaseEmulatedTest.mockFirestoreChangeObject(
      howtoApproved,
      howtoDraft,
      'howtos',
      howtoApproved._id,
    )

    await handleModerationUpdate(change, createHowtoModerationEmail)

    // No new emails should have been created
    const countSnapshot = await db.collection(DB_ENDPOINTS.emails).count().get()
    expect(countSnapshot.data().count).toEqual(1)
  })
})

describe('Create map pin moderation emails', () => {
  const db = FirebaseEmulatedTest.admin.firestore()

  beforeAll(async () => {
    await FirebaseEmulatedTest.clearFirestoreDB()
    await FirebaseEmulatedTest.seedFirestoreDB('emails')

    await FirebaseEmulatedTest.seedFirestoreDB('users', [
      userFactory('user_1', {
        displayName: 'User 1',
        userName: 'user_1',
        userRoles: ['beta-tester'],
      }),
      userFactory('user_2', {
        displayName: 'User 2',
        userName: 'user_2',
      }),
    ])

    await FirebaseEmulatedTest.seedFirestoreDB('mappins')
  })

  afterAll(async () => {
    await FirebaseEmulatedTest.clearFirestoreDB()
  })

  it('Creates an email for an accepted map pin', async () => {
    const mapPinApproved = {
      _id: 'map_pin_1',
      _createdBy: 'user_1',
      moderation: 'accepted',
    }
    const mapPinAwaitingModeration = {
      _id: 'map_pin_1',
      _createdBy: 'user_1',
      moderation: 'awaiting-moderation',
    }
    const change = FirebaseEmulatedTest.mockFirestoreChangeObject(
      mapPinAwaitingModeration,
      mapPinApproved,
      'mappins',
      mapPinApproved._id,
    )

    await handleModerationUpdate(change, createMapPinModerationEmail)

    // Only one approved howto email should have been created
    const countSnapshot = await db.collection(DB_ENDPOINTS.emails).count().get()
    expect(countSnapshot.data().count).toEqual(1)

    const querySnapshot = await db.collection(DB_ENDPOINTS.emails).get()
    querySnapshot.forEach((doc) => {
      const {
        message: { html, subject },
        to,
      } = doc.data()
      expect(html).toMatchSnapshot()
      expect(subject).toBe(MAP_PIN_APPROVAL_SUBJECT)
      // Check that the email contains the correct user name
      expect(html).toContain('Hey User 1')
      // Check that the email contains the correct map pin link
      expect(html).toContain(
        'https://community.preciousplastic.com/map#map_pin_1',
      )
      // Check that the email contains the correct PP signoff
      expect(html).toContain('Charlie your Precious Plastic Community Manager')
      expect(to).toBe('test@test.com')
    })
  })

  // Remove this test once released to all users.
  it('Does not creates email for people who are not beta testers', async () => {
    const mapPinApproved = {
      _id: 'map_pin_1',
      _createdBy: 'user_2',
      moderation: 'accepted',
    }
    const mapPinAwaitingModeration = {
      _id: 'map_pin_1',
      _createdBy: 'user_2',
      moderation: 'awaiting-moderation',
    }
    const change = FirebaseEmulatedTest.mockFirestoreChangeObject(
      mapPinAwaitingModeration,
      mapPinApproved,
      'mappins',
      mapPinApproved._id,
    )

    await handleModerationUpdate(change, createMapPinModerationEmail)

    // No new emails should have been created
    const countSnapshot = await db.collection(DB_ENDPOINTS.emails).count().get()
    expect(countSnapshot.data().count).toEqual(1)
  })

  it('Does not creates email for non-approved map pins', async () => {
    const mapPinDraft = {
      _id: 'map_pin_1',
      _createdBy: 'user_1',
      moderation: 'draft',
    }
    const mapPinAwaitingModeration = {
      _id: 'map_pin_1',
      _createdBy: 'user_1',
      moderation: 'awaiting-moderation',
    }
    const change = FirebaseEmulatedTest.mockFirestoreChangeObject(
      mapPinAwaitingModeration,
      mapPinDraft,
      'mappins',
      mapPinDraft._id,
    )

    await handleModerationUpdate(change, createMapPinModerationEmail)

    // No new emails should have been created
    const countSnapshot = await db.collection(DB_ENDPOINTS.emails).count().get()
    expect(countSnapshot.data().count).toEqual(1)
  })
})
