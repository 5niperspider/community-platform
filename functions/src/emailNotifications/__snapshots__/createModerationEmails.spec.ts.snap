// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create email test Creates an email for an accepted howto 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html>
  <head>
    <meta http-equiv="Content-Type" content="text/html charset=UTF-8" />
    <style>
      #body {
        font-family: Arial, Helvetica, sans-serif;
        font-size: 14px;
        color: #000000;
      }
      #body a {
        color: inherit;
      }
      .border {
        border: 3px solid black;
        border-radius: 10px;
        padding: 4% 0;
        margin: 4% 0;
        width: 550px;
      }
      table {
        width: 100%;
        border: 0;
      }
      .settings-table-container,
      .project-image-table-container {
        margin-bottom: 8%;
      }

      .greeting-container,
      .main-container {
        margin-bottom: 8%;
      }
      .greeting-container,
      .main-container,
      .closing-containter {
        margin-left: 8%;
      }
      @media only screen and (max-width: 550px) {
        .table-container {
          width: 100%;
        }
      }
    </style>
  </head>
  <body id="body">
    <table class="email-table-container">
      <tr>
        <td align="center">
          <div class="border">
            <table class="project-image-table-container">
              <tr>
                <td align="center">
                  <img
                    width="144"
                    alt=""
                    src="http://cdn.mcauto-images-production.sendgrid.net/cb5a685b085b7e7c/f48e2455-6dbb-4102-bde9-2ee07d3008c4/859x860.png"
                  />
                </td>
              </tr>
            </table>

            <div align="left" class="greeting-container">
              <p>Hey User 1</p>
              <p>Huzzah! Your How-To Mock Howto has been approved.</p>
            </div>
            <div align="left" class="main-container">
              <p>
                It is visible on the community platform
                <a
                  href="https://community.preciousplastic.com/how-to/00_user_1_howto"
                  >here</a
                >.
              </p>
              <p>Users are now be able to comment and mark it as useful.</p>
            </div>
            <div align="left" class="closing-containter">
              <p>Thanks for sharing your knowledge!</p>
              <p>Charlie your Precious Plastic Community Manager</p>
            </div>
          </div>
          <table class="settings-table-container">
            <tr>
              <td align="center">
                <div class="notifications">
                  Manage your notifications
                  <a href="https://community.preciousplastic.com/settings">
                    here</a
                  >
                </div>
              </td>
            </tr>
          </table>
        </td>
      </tr>
    </table>
  </body>
</html>
"
`;
